src/processing.rs

```diff
@@ -41,57 +41,81 @@ pub fn process_sign_message(
 
-    // If any of the inputs was already signed, return null
+    // Gather what signatures we have for these prevouts
+    let mut signatures = Vec::with_capacity(n_inputs);
     for txin in spend_tx.inner_tx().global.unsigned_tx.input.iter() {
-        if db_signed_outpoint(&db_path, &txin.previous_output)
+        if let Some(signed_op) = db_signed_outpoint(&db_path, &txin.previous_output)
             .map_err(SignProcessingError::Database)?
-            .is_some()
         {
-            return Ok(null_signature());
+            signatures.push(signed_op.signature)
         }
 
         // NOTE: we initially decided to check each manager's signature here, and then we discarded
         // it. This is still being discussed whether it's fine to drop this check...
+        // We later stripped the managers signatures from the PSBT before sharing it, in order to
+        // not hit the Noise message size limit. Hence the above is unlikely to happen anytime
+        // soon..
     }
 
-    // If we never signed it yet, append our signatures to the PSBT
-    let secp = secp256k1::Secp256k1::signing_only();
-    let our_pubkey = BitcoinPubkey {
-        compressed: true,
-        key: secp256k1::PublicKey::from_secret_key(&secp, bitcoin_privkey),
-    };
-    let mut psbtins = spend_tx.inner_tx_mut().inputs.clone(); // borrow checker forces a clone..
+    // If we had all the signatures for all these outpoints, send them. They'll figure out whether
+    // they are valid or not.
+    if signatures.len() == n_inputs {
+        for (i, sig) in signatures.into_iter().enumerate() {
+            spend_tx.inner_tx_mut().inputs[i]
+                .partial_sigs
+                .insert(our_pubkey, sig);
+        }
+        return Ok(SignResult { tx: Some(spend_tx) });
+    }
+
+    // If we already signed some of the outpoints, don't sign anything else!
+    if !signatures.is_empty() {
+        return Ok(null_signature());
+    }
+
+    // If we signed none of the input, append fresh signatures for each of them to the PSBT.
+    let mut psbtins = spend_tx.inner_tx().inputs.clone();
     for (i, psbtin) in psbtins.iter_mut().enumerate() {
         // FIXME: sighash cache upstream...
         let sighash = spend_tx
             .signature_hash_internal_input(i, SigHashType::All)
             .map_err(SignProcessingError::InsanePsbtMissingInput)?;
         let sighash = secp256k1::Message::from_slice(&sighash).expect("Sighash is 32 bytes");
-        let mut signature = secp
-            .sign(&sighash, bitcoin_privkey)
-            .serialize_der()
-            .to_vec();
-        signature.push(SigHashType::All as u8);
+
+        let signature = secp.sign(&sighash, bitcoin_privkey);
+        let mut raw_sig = signature.serialize_der().to_vec();
+        raw_sig.push(SigHashType::All as u8);
         assert!(
-            psbtin.partial_sigs.insert(our_pubkey, signature).is_none(),
+            psbtin.partial_sigs.insert(our_pubkey, raw_sig).is_none(),
```

Comment:
Not introduced in this PR but we should not crash there. They could just DOS us by sending a PSBT with our pubkey in the partial_sigs and a dummy sig

Reply:
Tracked in https://github.com/revault/cosignerd/issues/21